# Pandas DataFrames

Pandas DataFrames are mutable two-dimensional structures of data with labeled axes where: 
* each row represents a different observation
* each column represents a different variable

As always, we first need to import the Pandas module:

In [81]:
import pandas as pd

Next, if we want a DataFrame with 5 rows and 2 columns, we can do it from a [dictionary](https://www.w3schools.com/python/python_dictionaries.asp), a [list](https://www.w3schools.com/python/python_lists.asp) of lists, a list of dictionaries, etc.

We are going to create a 5 row, 2 column DataFrame from a dictionary. To do this, we will follow these steps:

1. Create a dictionary where the keys will be the names of the columns and the values will be lists, with as many elements as the number of rows we want.
2. Convert that dictionary to dataframe with pandas' `pd.DataFrame()` function.

In [4]:
# 1. Create the dictionary
data = {
    "x":[1, 2, 3, 4, 5], 
    "y":[6, 7, 8, 9, 10]
}

In [5]:
data['x']

[1, 2, 3, 4, 5]

In [6]:
# 2. Convert dictionary to dataframe
df = pd.DataFrame(data)
print(df)
# df

   x   y
0  1   6
1  2   7
2  3   8
3  4   9
4  5  10


As we said, we have created a dataframe with 5 rows and two columns, called x and y respectively.

**Observation**: As a result of `print()`, we have not only obtained the 5 rows and 2 columns, but there is an additional "column" of 5 numbers ordered vertically from 0 to 4. This column is called the **index** and it is simply the name of each row, which by default is their ordinal position: 0 indicates the first row; the 1, the second; and so on.

## Import a csv file to dataframe

Most of the time, you will not be creating dataframes yourself, but importing (or "reading") data from a csv file or a database into a pandas dataframe. It's easy to do with pandas' read functions:

In [82]:
url = 'https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing' # orderlines.csv
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
df = pd.read_csv(path)

## Dataframe dimensions

With the `.shape` attribute we can calculate the dimensions (number of rows and columns) of the dataframe.

In [83]:
df.shape

(293983, 7)

As a result we obtain a [tuple](https://www.w3schools.com/python/python_tuples.asp) where the first element is the number of rows, which in our case is 293983, while the second element is the number of columns, which in our example was 7.

In [11]:
nrows = df.shape[0]
ncols = df.shape[1]
print("The number of rows is", nrows)
print("The number of columns is", ncols)

The number of rows is 293983
The number of columns is 7


`DataFrame.size` returns the total number of values that the dataframe has (the number of rows multiplied by the number of columns):

In [14]:
df.size

2057881

In [15]:
# check if that's true
df.shape[0] * df.shape[1] == df.size

True

With the `.ndim` attribute we calculate the number of dimensions that the dataframe has. This will always be 2, as it consists of rows and columns.

In [17]:
df.ndim

2

## Dataframes exploration

The `DataFrame.head()` and `DataFrame.tail()` methods are used to display the first or last rows of the dataframe. Looking at the raw data is a great way to get a grasp of what's in there. By default, 5 rows will be shown, but you can change that:

In [18]:
df.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [19]:
df.head(9)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   
5  1119114    295310           0                10  WDT0249     231.79   
6  1119115    299544           0                 1  APP1582   1.137.99   
7  1119116    299545           0                 1  OWC0100      47.49   
8  1119119    299546           0                 1  IOT0014      18.99   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38  
5  2017-01-01 01:14:27  
6  2017-01-01 01:17:21  
7  2017-01-01 01:46:16  
8  2017-01-01 01:50:34

In [20]:
df.tail()

id  id_order  product_id  product_quantity      sku unit_price  \
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36

The methods `DataFrame.info()`, `DataFrame.describe()` and `DataFrame.nunique()` give a general overview of what's in the dataframe:

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


In [22]:
df.describe()

id       id_order  product_id  product_quantity
count  2.939830e+05  293983.000000    293983.0     293983.000000
mean   1.397918e+06  419999.116544         0.0          1.121126
std    1.530096e+05   66344.486479         0.0          3.396569
min    1.119109e+06  241319.000000         0.0          1.000000
25%    1.262542e+06  362258.500000         0.0          1.000000
50%    1.406940e+06  425956.000000         0.0          1.000000
75%    1.531322e+06  478657.000000         0.0          1.000000
max    1.650203e+06  527401.000000         0.0        999.000000

In [23]:
df.nunique()

id                  293983
id_order            204855
product_id               1
product_quantity        67
sku                   7951
unit_price           11329
date                251631
dtype: int64

The `.unique()` method returns the unique values from a column as a numpy array, which can be indexed with `[]`:

In [26]:
df['sku'].unique()[:10]

array(['OTT0133', 'LGE0043', 'PAR0071', 'WDT0315', 'JBL0104', 'WDT0249',
       'APP1582', 'OWC0100', 'IOT0014', 'APP0700'], dtype=object)

The `isna()` method returns a boolean for each value: `True` if that value is "missing" (which is represented as `NaN` in numpy and pandas) and `False` if the value is not missing:

In [28]:
df.isna()

id  id_order  product_id  product_quantity    sku  unit_price  \
0       False     False       False             False  False       False   
1       False     False       False             False  False       False   
2       False     False       False             False  False       False   
3       False     False       False             False  False       False   
4       False     False       False             False  False       False   
...       ...       ...         ...               ...    ...         ...   
293978  False     False       False             False  False       False   
293979  False     False       False             False  False       False   
293980  False     False       False             False  False       False   
293981  False     False       False             False  False       False   
293982  False     False       False             False  False       False   

         date  
0       False  
1       False  
2       False  
3       False  
4       False  
...       ...  
293978  False  
293979  False  
293980  False  
293981  False  
293982  False  

[293983 rows x 7 columns]

We can then use `DataFrame.sum()` to add up all these booleans for each column, and count how many missing values are there in the dataframe, since `True` is interpreted as `1` and `False` as `0`:

In [29]:
df.isna().sum()

id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64

`DataFrame.duplicated()` also returns a boolean output, but in this case just one value per row: `True` if that row is duplicated and `False` if it's not. Again, using `sum()` allows us to count how many `True` values (i.e. duplicated rows) are there in total:

In [30]:
df.duplicated().sum() # parameters keep=False
# df.drop_duplicates()

0

`DataFrame.nlargest(n, columns)` will return the top `n` rows with the largest value for whatever column we specify in `columns`. Below, we see the rows with the largest product quantity values:

In [31]:
df.nlargest(5, 'product_quantity')

id  id_order  product_id  product_quantity      sku unit_price  \
53860  1228150    346221           0               999  APP1190      55.99   
68712  1254032    358747           0               999  SEV0028      19.99   
57796  1234924    349475           0               800  KIN0137       7.49   
57306  1234111    349133           0               555  APP0665      70.99   
40813  1204788    335057           0               201  THU0029      80.99   

                      date  
53860  2017-04-14 21:50:52  
68712  2017-05-24 14:51:58  
57796  2017-04-25 09:59:00  
57306  2017-04-24 10:20:13  
40813  2017-03-14 15:25:53

 `DataFrame.nsmallest()` does the same, for the smallest values:

In [32]:
df.nsmallest(5, 'product_quantity')

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

So far, we have seen two ways to explore DataFrames:

* `.shape`, `.size` and `.ndim`, among others, are **atributes**. They are written without parentheses and give you raw "metadata" about the DataFrame you are calling them on.
* `.head()`, `.describe()` and `.isna()`, among others, are **methods**. They are written with parentheses and perform some sort of calculation, transformation or aggregation. A method is like a function that is tied to a specific object type.

DataFrames have a lot of attributes and methods and may not be obvious whether something belongs to one type or the other. Whenever in doubt, check [the documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html?highlight=dataframe#pandas.DataFrame). If you scroll down past the list of examples you will find two sections listing all the attributes and methods.

## Columns

Given a DataFrame, we can select a particular column in several ways:

* Indicating the name of the column between square brackets, `[]`
* With the `.loc[]` attribute (by name or tag)
* With the `.iloc[]` attribute (by position)

Plain square brackets `[]` are used to just view a column if you know its name and **don't want to modify it**:

In [33]:
# select the column by name
df['id_order']

0         299539
1         299540
2         299541
3         299542
4         299543
           ...  
293978    527398
293979    527399
293980    527400
293981    527388
293982    527401
Name: id_order, Length: 293983, dtype: int64

`.loc[]` takes two arguments: `[rows, columns]`. Passing `:` to the rows argument means "grabbing all the rows", which allows you to select a whole column if you know its name. This method is more flexible as you will see in the future, and allows you to modify the data.

In [34]:
# method .loc[]
df.loc[:, 'id_order']

0         299539
1         299540
2         299541
3         299542
4         299543
           ...  
293978    527398
293979    527399
293980    527400
293981    527388
293982    527401
Name: id_order, Length: 293983, dtype: int64

`iloc[]` works similarly, but only accepts integers (which represent the positions of the rows and columns):

In [35]:
# method .iloc[]
print(df.iloc[:, 0])

0         1119109
1         1119110
2         1119111
3         1119112
4         1119113
           ...   
293978    1650199
293979    1650200
293980    1650201
293981    1650202
293982    1650203
Name: id, Length: 293983, dtype: int64


### Select multiple columns

If we want to select more than one column, we can do it with all the options listed above, with slight modifications in some cases:

In [36]:
# note that we pass a list inside of the []
df[['id_order','sku']]

id_order      sku
0         299539  OTT0133
1         299540  LGE0043
2         299541  PAR0071
3         299542  WDT0315
4         299543  JBL0104
...          ...      ...
293978    527398  JBL0122
293979    527399  PAC0653
293980    527400  APP0698
293981    527388  BEZ0204
293982    527401  APP0927

[293983 rows x 2 columns]

In [37]:
# .loc()
df.loc[:, ["id_order", "sku"]]

id_order      sku
0         299539  OTT0133
1         299540  LGE0043
2         299541  PAR0071
3         299542  WDT0315
4         299543  JBL0104
...          ...      ...
293978    527398  JBL0122
293979    527399  PAC0653
293980    527400  APP0698
293981    527388  BEZ0204
293982    527401  APP0927

[293983 rows x 2 columns]

With `.loc[]` and `:` you can select all columns between two columns you specify.

In [38]:
# .loc()
df.loc[:, "id_order":"sku"]

id_order  product_id  product_quantity      sku
0         299539           0                 1  OTT0133
1         299540           0                 1  LGE0043
2         299541           0                 1  PAR0071
3         299542           0                 1  WDT0315
4         299543           0                 1  JBL0104
...          ...         ...               ...      ...
293978    527398           0                 1  JBL0122
293979    527399           0                 1  PAC0653
293980    527400           0                 2  APP0698
293981    527388           0                 1  BEZ0204
293982    527401           0                 1  APP0927

[293983 rows x 4 columns]

In [39]:
# .iloc
df.iloc[:, [0, 1]]

id  id_order
0       1119109    299539
1       1119110    299540
2       1119111    299541
3       1119112    299542
4       1119113    299543
...         ...       ...
293978  1650199    527398
293979  1650200    527399
293980  1650201    527400
293981  1650202    527388
293982  1650203    527401

[293983 rows x 2 columns]

In [40]:
df.iloc[:, 0:2]

id  id_order
0       1119109    299539
1       1119110    299540
2       1119111    299541
3       1119112    299542
4       1119113    299543
...         ...       ...
293978  1650199    527398
293979  1650200    527399
293980  1650201    527400
293981  1650202    527388
293982  1650203    527401

[293983 rows x 2 columns]

## Rows

Selecting rows is easy if you know how to select columns. You have two options:

* With `.loc[]` (by name or tag)
* With `.iloc[]` (by position)

Selecting a single row returns a pandas Series (the 1-dimensional object that pandas has):

In [41]:
df.loc[0]

id                              1119109
id_order                         299539
product_id                            0
product_quantity                      1
sku                             OTT0133
unit_price                        18.99
date                2017-01-01 00:07:19
Name: 0, dtype: object

With `.loc[]`, rows are selected by its index name:

In [42]:
df.loc[0:3]

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40

If we change the index and set it to the `id` column, now the first rows can not be selected the same way:

In [43]:
df.set_index('id', inplace = True)
# inplace = True is the same as doing df = df.set_index('id'), i.e. it modifies the dataframe

In [45]:
df.loc[0:3]

Empty DataFrame
Columns: [id_order, product_id, product_quantity, sku, unit_price, date]
Index: []

In [46]:
df.head(2)

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119109    299539           0                 1  OTT0133      18.99   
1119110    299540           0                 1  LGE0043     399.00   

                        date  
id                            
1119109  2017-01-01 00:07:19  
1119110  2017-01-01 00:19:45

In [47]:
# select the first observation with the .loc() method
df.loc[1119110]

id_order                         299540
product_id                            0
product_quantity                      1
sku                             LGE0043
unit_price                       399.00
date                2017-01-01 00:19:45
Name: 1119110, dtype: object

With the `iloc[]` method we don't need to know the row names to select rows at a certain position:

In [48]:
df.iloc[0:3]

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119109    299539           0                 1  OTT0133      18.99   
1119110    299540           0                 1  LGE0043     399.00   
1119111    299541           0                 1  PAR0071     474.05   

                        date  
id                            
1119109  2017-01-01 00:07:19  
1119110  2017-01-01 00:19:45  
1119111  2017-01-01 00:20:57

In [49]:
# select the last observation with the method .iloc[]
df.iloc[-1]
# df.tail(1)

id_order                         527401
product_id                            0
product_quantity                      1
sku                             APP0927
unit_price                        13.99
date                2018-03-14 13:58:36
Name: 1650203, dtype: object

In [50]:
df.loc[[1119111,1119112,1119113]]

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119111    299541           0                 1  PAR0071     474.05   
1119112    299542           0                 1  WDT0315      68.39   
1119113    299543           0                 1  JBL0104      23.74   

                        date  
id                            
1119111  2017-01-01 00:20:57  
1119112  2017-01-01 00:51:40  
1119113  2017-01-01 01:06:38

Again, using `:` you can select all elements between the two that you specified:

In [51]:
df.loc[1119111:1119113]

id_order  product_id  product_quantity      sku unit_price  \
id                                                                    
1119111    299541           0                 1  PAR0071     474.05   
1119112    299542           0                 1  WDT0315      68.39   
1119113    299543           0                 1  JBL0104      23.74   

                        date  
id                            
1119111  2017-01-01 00:20:57  
1119112  2017-01-01 00:51:40  
1119113  2017-01-01 01:06:38

In [69]:
df.reset_index(inplace = True)

Indexing can get tricky sometimes, it's ok to take some time to get used to the methods we presented, and it's ok to have some trouble selecting the rows and columns you need. For an exhaustive guide on Pandas indexing, check out this link: https://pandas.pydata.org/docs/user_guide/indexing.html#indexing 

## Drop and Filter data

The `.drop()` method allows us to delete the rows or columns that we indicate.

**Note:** Again, if we want to directly apply the changes to the original dataframe, we need to indicate `inplace = True`. Otherwise, we are getting as an output just a "view" of how the dataframe looks like after the drop, but the original dataframe remains untouched.

In [70]:
# droping the column "unit_price"
# axis=1 means we want to drop a column, not a row (for rows, axis=0)
df.drop(['unit_price'], axis = 1)

index       id  id_order  product_id  product_quantity      sku  \
0            0  1119109    299539           0                 1  OTT0133   
1            1  1119110    299540           0                 1  LGE0043   
2            2  1119111    299541           0                 1  PAR0071   
3            3  1119112    299542           0                 1  WDT0315   
4            4  1119113    299543           0                 1  JBL0104   
...        ...      ...       ...         ...               ...      ...   
293978  293978  1650199    527398           0                 1  JBL0122   
293979  293979  1650200    527399           0                 1  PAC0653   
293980  293980  1650201    527400           0                 2  APP0698   
293981  293981  1650202    527388           0                 1  BEZ0204   
293982  293982  1650203    527401           0                 1  APP0927   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 7 columns]

How to filter information on a dataframe.

In [72]:
# rows with product quantity larger than 100
df.loc[df['product_quantity'] > 100, :]

index       id  id_order  product_id  product_quantity      sku  \
27779    27779  1180010    323959           0               126  ADN0039   
40813    40813  1204788    335057           0               201  THU0029   
53860    53860  1228150    346221           0               999  APP1190   
57306    57306  1234111    349133           0               555  APP0665   
57796    57796  1234924    349475           0               800  KIN0137   
68712    68712  1254032    358747           0               999  SEV0028   
136675  136675  1388261    417536           0               200  TRK0009   
204637  204637  1500715    464858           0               192  APP1662   
246048  246048  1574262    496172           0               164  EVU0013   
285492  285492  1637611    522075           0               125  XDO0047   

       unit_price                 date  
27779       34.99  2017-02-14 10:21:12  
40813       80.99  2017-03-14 15:25:53  
53860       55.99  2017-04-14 21:50:52  
57306       70.99  2017-04-24 10:20:13  
57796        7.49  2017-04-25 09:59:00  
68712       19.99  2017-05-24 14:51:58  
136675      29.99  2017-10-25 15:02:39  
204637     519.00  2017-12-17 15:53:04  
246048      19.99  2018-01-22 16:14:42  
285492      25.99  2018-03-06 10:07:54

The `.query()` method can be useful for this purpose, as it resembles SQL syntax. Note that it works only when the column values do not contain blank spaces. You can use any **Python Comparison Operators** you want inside the query method (find more information on this [link](https://www.w3schools.com/python/python_operators.asp)).

In [73]:
df.query('product_quantity > 100')

index       id  id_order  product_id  product_quantity      sku  \
27779    27779  1180010    323959           0               126  ADN0039   
40813    40813  1204788    335057           0               201  THU0029   
53860    53860  1228150    346221           0               999  APP1190   
57306    57306  1234111    349133           0               555  APP0665   
57796    57796  1234924    349475           0               800  KIN0137   
68712    68712  1254032    358747           0               999  SEV0028   
136675  136675  1388261    417536           0               200  TRK0009   
204637  204637  1500715    464858           0               192  APP1662   
246048  246048  1574262    496172           0               164  EVU0013   
285492  285492  1637611    522075           0               125  XDO0047   

       unit_price                 date  
27779       34.99  2017-02-14 10:21:12  
40813       80.99  2017-03-14 15:25:53  
53860       55.99  2017-04-14 21:50:52  
57306       70.99  2017-04-24 10:20:13  
57796        7.49  2017-04-25 09:59:00  
68712       19.99  2017-05-24 14:51:58  
136675      29.99  2017-10-25 15:02:39  
204637     519.00  2017-12-17 15:53:04  
246048      19.99  2018-01-22 16:14:42  
285492      25.99  2018-03-06 10:07:54

The `isin()` method is very useful to find rows that match any of the values you have in a list. For example, here we are searching for rows where its `sku` matches any of the 2 sku's we listed:

In [74]:
# find out a column that contains a list
df['sku'].isin(['JBL0104', 'ADN0039'])

0         False
1         False
2         False
3         False
4          True
          ...  
293978    False
293979    False
293980    False
293981    False
293982    False
Name: sku, Length: 293983, dtype: bool

This expression can be used inside of `[]` or `.loc[]` to filter the rows that have a `True` value. This is called "boolean indexing" and it is really useful:

In [75]:
my_products = df['sku'].isin(['JBL0104', 'ADN0039'])

df.loc[my_products, :]

index       id  id_order  product_id  product_quantity      sku  \
4            4  1119113    299543           0                 1  JBL0104   
716        716  1120463    300212           0                 2  JBL0104   
774        774  1120607    300272           0                 1  JBL0104   
866        866  1120758    300349           0                 1  JBL0104   
1411      1411  1122014    300851           0                 1  JBL0104   
...        ...      ...       ...         ...               ...      ...   
258180  258180  1593360    503247           0                 1  JBL0104   
263790  263790  1601885    506553           0                 1  JBL0104   
266635  266635  1606326    508345           0                 2  JBL0104   
268446  268446  1608949    509377           0                 1  JBL0104   
279801  279801  1628218    517916           0                 1  JBL0104   

       unit_price                 date  
4           23.74  2017-01-01 01:06:38  
716         26.99  2017-01-02 01:44:05  
774         24.99  2017-01-02 08:57:07  
866         24.99  2017-01-02 10:41:29  
1411        24.99  2017-01-02 18:50:13  
...           ...                  ...  
258180      23.99  2018-01-31 20:55:50  
263790      24.99  2018-02-05 22:41:40  
266635      24.99  2018-02-08 19:20:21  
268446      24.99  2018-02-10 18:02:37  
279801      27.99  2018-02-26 20:51:30  

[65 rows x 8 columns]

Pandas compresses large outputs in Colab/Jupyter Notebooks. If you want to see more rows, you can change the options:

In [ ]:
pd.options.display.max_rows = 100
# after running this cell, run the code above again and you will see all the rows.

For a complete list of all settings and options that can be tweaked, check out this: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

#### Modifying a dataframe & the `.copy()` method

Let's take a small chunk of data from our dataframe:

In [84]:
sample = df.iloc[:3,:]
sample

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

Now we pick a single cell from the sample we took and we assign a new value to it. A warning already tells us that this is a risky thing to do:

In [87]:
# we change the "unit_price" of the first row:
sample.iloc[0,4] = 'NEW VALUE HERE'

We can see the new value on the `sample` we took:

In [88]:
sample

id  id_order  product_id  product_quantity             sku unit_price  \
0  1119109    299539           0                 1  NEW VALUE HERE      18.99   
1  1119110    299540           0                 1         LGE0043     399.00   
2  1119111    299541           0                 1         PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

..and, maybe to your surprise, we can see that the new value is also present on the original `df`!

In [89]:
df.head()

id  id_order  product_id  product_quantity             sku unit_price  \
0  1119109    299539           0                 1  NEW VALUE HERE      18.99   
1  1119110    299540           0                 1         LGE0043     399.00   
2  1119111    299541           0                 1         PAR0071     474.05   
3  1119112    299542           0                 1         WDT0315      68.39   
4  1119113    299543           0                 1         JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

When you take a chunk of data using `.loc[]` or `iloc[]` and assign it to a new object, the new object is just a "tag" pointing to the very same data as the original dataframe points to. We can avoid this using the method `.copy()`

In [91]:
url = 'https://drive.google.com/file/d/14Y7g5ITyf6LMyPoKc9wr010V9StaCUux/view?usp=sharing' # orderlines.csv
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
df = pd.read_csv(path)

sample = df.iloc[:3,:].copy()
sample.iloc[0,4] = 'NEW VALUE HERE'
sample

id  id_order  product_id  product_quantity             sku unit_price  \
0  1119109    299539           0                 1  NEW VALUE HERE      18.99   
1  1119110    299540           0                 1         LGE0043     399.00   
2  1119111    299541           0                 1         PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

In [93]:
df.head(3)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57

As you can see, now it has not been modified.

# CHALLENGES

In [2]:
import pandas as pd

In [3]:
orderlines = pd.read_csv("orderlines.csv")
orderlines.sample(10)

id  id_order  product_id  product_quantity      sku unit_price  \
149620  1414262    428798           0                 1  APP1922     154.00   
2834    1128018    302109           0                 1  WOE0009      12.99   
273353  1617449    513202           0                 1  OWC0013       2.99   
265498  1604614    507670           0                 1  MIC0036      68.99   
285471  1637574    522055           0                 1  IFX0171     119.99   
192002  1482137    457263           0                 1  APP1190      56.00   
155285  1425547    431683           0                 1  GRT0434      14.44   
38112   1199611    332793           0                 1  LGE0048     244.99   
231962  1550798    487087           0                 1  APP1190      54.00   
41544   1206132    335719           0                 1  NTE0063     109.99   

                       date  
149620  2017-11-20 00:23:57  
2834    2017-01-03 22:53:31  
273353  2018-02-17 09:44:01  
265498  2018-02-07 18:20:54  
285471  2018-03-06 09:10:33  
192002  2017-12-07 17:48:56  
155285  2017-11-23 08:50:19  
38112   2017-03-07 00:18:35  
231962  2018-01-09 22:28:31  
41544   2017-03-16 10:39:02

In [4]:
orderlines.duplicated().sum()

0

1. How many different unit prices does the product with the sku JBL0104 have? Combine a pandas filter method with the method `.nunique()`.

In [5]:
orderlines[orderlines["sku"] == "JBL0104"]

id  id_order  product_id  product_quantity      sku unit_price  \
4       1119113    299543           0                 1  JBL0104      23.74   
716     1120463    300212           0                 2  JBL0104      26.99   
774     1120607    300272           0                 1  JBL0104      24.99   
866     1120758    300349           0                 1  JBL0104      24.99   
1411    1122014    300851           0                 1  JBL0104      24.99   
1451    1122093    300887           0                 1  JBL0104      24.99   
3612    1130604    302848           0                 1  JBL0104      24.99   
4370    1132185    303586           0                 1  JBL0104      22.31   
8450    1140943    307256           0                 1  JBL0104      26.99   
14439   1153300    312363           0                 1  JBL0104      26.99   
53150   1226898    345622           0                 1  JBL0104      26.99   
56740   1233192    348682           0                 1  JBL0104      26.99   
98730   1307785    384545           0                 1  JBL0104      26.99   
105968  1320675    390750           0                 1  JBL0104      26.99   
112369  1344362    396145           0                 1  JBL0104      26.99   
154201  1423772    432828           0                 1  JBL0104      22.99   
165640  1443880    442197           0                 1  JBL0104      22.99   
165688  1443963    442228           0                 1  JBL0104      22.99   
171548  1453081    446003           0                 1  JBL0104      22.99   
172402  1454757    446688           0                 1  JBL0104      22.99   
192427  1482783    457514           0                 1  JBL0104      24.99   
195671  1486865    459056           0                 1  JBL0104      24.99   
196563  1488065    459522           0                 1  JBL0104      24.99   
203901  1499365    464210           0                 1  JBL0104      24.99   
205833  1503035    465885           0                 1  JBL0104      24.99   
206138  1503671    466170           0                 1  JBL0104      24.99   
206312  1503978    466344           0                 1  JBL0104      24.99   
206905  1505114    466844           0                 1  JBL0104      24.99   
207094  1505489    467005           0                 1  JBL0104      24.99   
207544  1506336    467370           0                 1  JBL0104      24.99   
211840  1514696    471338           0                 1  JBL0104      23.74   
212820  1516518    472215           0                 1  JBL0104      23.74   
213153  1517116    472506           0                 1  JBL0104      23.74   
213262  1517297    472597           0                 1  JBL0104      23.74   
213808  1518224    473037           0                 1  JBL0104      23.74   
213841  1518286    473074           0                 1  JBL0104      24.99   
214057  1518672    473261           0                 1  JBL0104      23.74   
214559  1519654    473732           0                 1  JBL0104      23.74   
216307  1522981    475353           0                 1  JBL0104      23.74   
217798  1525891    476696           0                 1  JBL0104      23.74   
218027  1526318    476911           0                 1  JBL0104      23.74   
219284  1528937    478069           0                 1  JBL0104      23.74   
220630  1531600    479235           0                 1  JBL0104      23.74   
221210  1532764    479755           0                 1  JBL0104      23.74   
221216  1532774    479762           0                 1  JBL0104      23.74   
222468  1535257    480924           0                 1  JBL0104      24.99   
231934  1550720    487057           0                 1  JBL0104      23.99   
233400  1553253    487975           0                 1  JBL0104      23.99   
234637  1555195    480555           0                 2  JBL0104      23.99   
234895  1555641    488957           0                 1  JBL010

In [6]:
orderlines[orderlines.sku == "JBL0104"]["unit_price"].nunique()

7

In [7]:
orderlines[orderlines.sku == "JBL0104"].unit_price.nunique()

7

In [8]:
orderlines.query("sku == 'JBL0104'").unit_price.nunique()

7

2. List the (unique) items that were sold in the order with the id_order 385921.

In [15]:
orderlines[orderlines.id_order == 385921].sku.unique()

array(['APP2431', 'APP2348', 'APP2131', 'APP1630', 'APP1735', 'APP1216',
       'APP2092', 'APP1215', 'ELA0017', 'MIN0010', 'ELA0039', 'BEA0046',
       'BOS0034', 'BEA0071', 'ELA0029', 'APP2161', 'HOC0008', 'NOM0026',
       'NOM0014'], dtype=object)

In [17]:
orderlines.query("id_order == 385921").sku.unique().tolist()

['APP2431',
 'APP2348',
 'APP2131',
 'APP1630',
 'APP1735',
 'APP1216',
 'APP2092',
 'APP1215',
 'ELA0017',
 'MIN0010',
 'ELA0039',
 'BEA0046',
 'BOS0034',
 'BEA0071',
 'ELA0029',
 'APP2161',
 'HOC0008',
 'NOM0026',
 'NOM0014']

3. Consider the products with the sku's APP2431 and APP2348. Find out in how many orders were they present.

In [203]:
orderlines[(orderlines.sku == "APP2431") | (orderlines.sku == "APP2348")].id_order.count()

189

In [205]:
set1 = set(orderlines["id_order"][orderlines.sku == "APP2431"])

In [206]:
set2 = set(orderlines["id_order"][orderlines.sku == "APP2348"])

In [198]:
len(set1.intersection(set2))

10

In [18]:
orderlines[orderlines["sku"].isin(["APP2431", "APP2348"])].id_order.nunique()

179

In [19]:
orderlines[orderlines.sku.isin(["APP2431", "APP2348"])].id_order.nunique()

179

4. Create a new dataframe, `df_50`, with all the rows that have a product quantity higher than 500 and only the columns `id`, id `order`, `product_quantity` and `sku`. Be sure to use the method `.copy()`. Once the new dataframe is created, modify the column 'product_quantity' to 'quantity', and 'sku' to 'product_code'. To do so, you can use the method `.rename()` or assign a list of new names to the attribute `.columns`.

In [215]:
df_50 = orderlines[["id", "id_order", "product_quantity", "sku"]][orderlines.product_quantity > 500].copy()
df_50.columns = ["id", "id_order", "quantity", "product_code"]
df_50

id  id_order  quantity product_code
53860  1228150    346221       999      APP1190
57306  1234111    349133       555      APP0665
57796  1234924    349475       800      KIN0137
68712  1254032    358747       999      SEV0028

5. Filter all the order lines where the product with the sku XDO0047 has appeared. Sort the rows by their product quantity using the [`.sort_values()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html) method from pandas, in a DESCENDING order. Then look at the main descriptive information of this results with the method `.describe()`.

In [20]:
orderlines[orderlines.sku == "XDO0047"].product_quantity.sort_values(ascending = False).describe()

count     36.000000
mean       4.472222
std       20.662576
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max      125.000000
Name: product_quantity, dtype: float64

In [22]:
orderlines[orderlines.sku == "XDO0047"].sort_values("product_quantity", ascending = False).describe()

id       id_order  product_id  product_quantity
count  3.600000e+01      36.000000        36.0         36.000000
mean   1.485767e+06  458640.277778         0.0          4.472222
std    8.770532e+04   36891.598431         0.0         20.662576
min    1.365515e+06  406387.000000         0.0          1.000000
25%    1.405830e+06  425462.750000         0.0          1.000000
50%    1.471428e+06  453247.000000         0.0          1.000000
75%    1.564540e+06  492406.000000         0.0          1.000000
max    1.640821e+06  523533.000000         0.0        125.000000